index.html

```diff
@@ -816,36 +816,56 @@
           </dt>
         </dl>
         <p>
-          Upon receiving a <a>push message</a> for a <a>push subscription</a> from the <a>push
-          service</a> the <a>user agent</a> MUST run the following steps:
+          Upon receiving a <a>push message</a> from the <a>push service</a> the <a>user agent</a>
+          MUST run the following steps:
         </p>
         <ol>
-          <li>If the <a>Service Worker</a> associated with the <a>webapp</a> is not running, start
-          it.
-          </li>
-          <li>Let <var>scope</var> be the <code>ServiceWorkerGlobalScope</code> of the <a>Service
-          Worker</a> associated with the <a>webapp</a>.
+          <li>Let <var>registration</var> be the <a>service worker registration</a> corresponding
+          to the push message.
           </li>
-          <li>Let <var>event</var> be a new <a>PushEvent</a>, whose <code id="widl-PushEvent-data">
-            data</code> attribute is a new <a>PushMessageData</a> with <a>bytes</a> set to the
-            binary message data received by the <a>user agent</a> in the <a>push message</a>, or an
-            empty byte sequence if no data was received.
+          <li>If <var>registration</var> is not found, abort these steps.
           </li>
-          <li>
-            <a>Queue a task</a> to <a data-lt="fire a simple event">fire <var>event</var> as a
-            simple event</a> named <code>push</code> at <var>scope</var>.
+          <li>Invoke the <a>Handle Functional Event</a> algorithm with a <a>service worker
+          registration</a> of <var>registration</var> and <var>callbackSteps</var> set to the
+          following steps:
+            <ol>
+              <li>Set <var>global</var> to the global object that was provided as an argument.
+              </li>
+              <li>Create a <a>trusted event</a>, <var>e</var>, that uses the
+              <code><a>PushMessageEvent</a></code> interface, with the event type
+              <code>push</code>, which does not bubble, is not cancelable, and has no default
+              action.
+              </li>
+              <li>If the <a>push message</a> contains valid data, set the <code>data</code>
```

Comment:
What does _valid data_ mean? I don't want to fire `onpush` for messages that have undecryptable payloads.